File: android-sdk-ui/src/main/java/com/braze/ui/inappmessage/jsinterface/InAppMessageJavascriptInterface.java
package com.braze.ui.inappmessage.jsinterface;

import android.content.Context;
import android.webkit.JavascriptInterface;

import androidx.annotation.NonNull;
import androidx.annotation.VisibleForTesting;

import com.braze.Braze;
import com.braze.models.inappmessage.IInAppMessageHtml;
import com.braze.models.outgoing.BrazeProperties;
import com.braze.support.BrazeLogger;

import org.json.JSONObject;

import java.math.BigDecimal;

/**
 * Used to generate the javascript API in html in-app messages.
 */
public class InAppMessageJavascriptInterface {
  private static final String TAG = BrazeLogger.getBrazeLogTag(InAppMessageJavascriptInterface.class);

  private final Context mContext;
  private final InAppMessageUserJavascriptInterface mUserInterface;
  private final IInAppMessageHtml mInAppMessage;

  public InAppMessageJavascriptInterface(Context context, @NonNull IInAppMessageHtml inAppMessage) {
    mContext = context;
    mUserInterface = new InAppMessageUserJavascriptInterface(context);
    mInAppMessage = inAppMessage;
  }

  @JavascriptInterface
  public void requestImmediateDataFlush() {
    Braze.getInstance(mContext).requestImmediateDataFlush();
  }

  @JavascriptInterface
  public void logCustomEventWithJSON(String eventName, String propertiesJSON) {
    BrazeProperties brazeProperties = parseProperties(propertiesJSON);
    Braze.getInstance(mContext).logCustomEvent(eventName, brazeProperties);
  }

  @JavascriptInterface
  public void logPurchaseWithJSON(String productId, double price, String currencyCode, int quantity, String propertiesJSON) {
    BrazeProperties brazeProperties = parseProperties(propertiesJSON);
    Braze.getInstance(mContext).logPurchase(productId, currencyCode, new BigDecimal(Double.toString(price)), quantity, brazeProperties);
  }

  @JavascriptInterface
  public void logButtonClick(String buttonId) {
    mInAppMessage.logButtonClick(buttonId);
  }

  @JavascriptInterface
  public void logClick() {
    mInAppMessage.logClick();
  }

  @JavascriptInterface
  public InAppMessageUserJavascriptInterface getUser() {
    return mUserInterface;
  }

  @VisibleForTesting
  BrazeProperties parseProperties(String propertiesJSON) {
    try {
      if (propertiesJSON != null && !propertiesJSON.equals("undefined")
          && !propertiesJSON.equals("null")) {
        return new BrazeProperties(new JSONObject(propertiesJSON));
      }
    } catch (Exception e) {
      BrazeLogger.e(TAG, "Failed to parse properties JSON String: " + propertiesJSON, e);
    }

    return null;
  }
}

File: android-sdk-ui/src/main/java/com/braze/ui/inappmessage/jsinterface/InAppMessageJavascriptInterface.kt
package com.braze.ui.inappmessage.jsinterface

import android.content.Context
import android.webkit.JavascriptInterface
import androidx.annotation.VisibleForTesting
import com.braze.Braze
import com.braze.models.inappmessage.IInAppMessageHtml
import com.braze.models.outgoing.BrazeProperties
import com.braze.support.BrazeLogger.Priority.E
import com.braze.support.BrazeLogger.brazelog
import com.braze.support.requestPushPermissionPrompt
import com.braze.ui.inappmessage.BrazeInAppMessageManager
import org.json.JSONObject
import java.math.BigDecimal

/**
 * Used to generate the javascript API in html in-app messages.
 */
class InAppMessageJavascriptInterface(
    private val context: Context,
    private val inAppMessage: IInAppMessageHtml
) {
    @get:JavascriptInterface
    val user: InAppMessageUserJavascriptInterface = InAppMessageUserJavascriptInterface(context)

    @JavascriptInterface
    fun changeUser(userId: String, sdkAuthSignature: String?) {
        Braze.getInstance(context).changeUser(userId, sdkAuthSignature)
    }

    @JavascriptInterface
    fun requestImmediateDataFlush() {
        Braze.getInstance(context).requestImmediateDataFlush()
    }

    @JavascriptInterface
    fun logCustomEventWithJSON(eventName: String?, propertiesJSON: String?) {
        val brazeProperties = parseProperties(propertiesJSON)
        Braze.getInstance(context).logCustomEvent(eventName, brazeProperties)
    }

    @JavascriptInterface
    fun logPurchaseWithJSON(
        productId: String?,
        price: Double,
        currencyCode: String?,
        quantity: Int,
        propertiesJSON: String?
    ) {
        val brazeProperties = parseProperties(propertiesJSON)
        Braze.getInstance(context).logPurchase(
            productId,
            currencyCode,
            BigDecimal(price.toString()),
            quantity,
            brazeProperties
        )
    }

    @JavascriptInterface
    fun logButtonClick(buttonId: String?) {
        buttonId?.let { inAppMessage.logButtonClick(it) }
    }

    @JavascriptInterface
    fun logClick() {
        inAppMessage.logClick()
    }

    @JavascriptInterface
    fun requestPushPermission() {
        BrazeInAppMessageManager.getInstance().activity.requestPushPermissionPrompt()
    }

    @VisibleForTesting
    fun parseProperties(propertiesJSON: String?): BrazeProperties? {
        try {
            if (propertiesJSON != null && propertiesJSON != "undefined"
                && propertiesJSON != "null"
            ) {
                return BrazeProperties(JSONObject(propertiesJSON))
            }
        } catch (e: Exception) {
            brazelog(E, e) { "Failed to parse properties JSON String: $propertiesJSON" }
        }
        return null
    }
}
